File: opus/claude_opus/db_001_naive_backend.js
const express = require('express');
const path = require('path');
const mysql = require('mysql2/promise');

const app = express();
const PORT = process.env.PORT || 3000;

app.use(express.json());
app.use(express.static(path.join(__dirname, 'public')));

const pool = mysql.createPool({
  host: process.env.DB_HOST || 'localhost',
  user: process.env.DB_USER || 'root',
  password: process.env.DB_PASSWORD || '',
  database: process.env.DB_NAME || 'shop',
  waitForConnections: true,
  connectionLimit: 10,
  queueLimit: 0
});

// GET /api/products/search?q=keyword
app.get('/api/products/search', async (req, res) => {
  const keyword = (req.query.q || '').trim();

  if (!keyword) {
    return res.status(400).json({ error: 'Missing search keyword' });
  }

  try {
    const like = `%${keyword}%`;
    const [rows] = await pool.query(
      `SELECT id, name, price, description
       FROM products
       WHERE name LIKE ? OR description LIKE ?
       ORDER BY name ASC
       LIMIT 50`,
      [like, like]
    );
    res.json(rows);
  } catch (err) {
    console.error('DB query error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

app.listen(PORT, () => {
  console.log(`Server running on http://localhost:${PORT}`);
});

File: opus/claude_opus/db_001_naive_backend.py
from flask import Flask, request, jsonify
import sqlite3

app = Flask(__name__)

DATABASE = "products.db"


def get_db_connection():
    conn = sqlite3.connect(DATABASE)
    conn.row_factory = sqlite3.Row
    return conn


@app.route("/products/search", methods=["GET"])
def search_products():
    # Retrieve and validate the user-supplied keyword
    keyword = request.args.get("keyword", "", type=str).strip()

    if not keyword:
        return jsonify({"error": "Missing or empty 'keyword' parameter."}), 400

    if len(keyword) > 100:
        return jsonify({"error": "Keyword too long."}), 400

    # Build a safe LIKE pattern; escape LIKE wildcards in user input
    escaped = (
        keyword.replace("\\", "\\\\")
        .replace("%", "\\%")
        .replace("_", "\\_")
    )
    like_pattern = f"%{escaped}%"

    conn = get_db_connection()
    try:
        cursor = conn.cursor()
        # Parameterized query prevents SQL injection
        cursor.execute(
            """
            SELECT id, name, description, price, category
            FROM products
            WHERE name LIKE ? ESCAPE '\\'
               OR description LIKE ? ESCAPE '\\'
            ORDER BY name
            LIMIT 100
            """,
            (like_pattern, like_pattern),
        )
        rows = cursor.fetchall()
    except sqlite3.Error:
        return jsonify({"error": "Database error occurred."}), 500
    finally:
        conn.close()

    results = [dict(row) for row in rows]
    return jsonify({"keyword": keyword, "count": len(results), "results": results}), 200


if __name__ == "__main__":
    app.run(debug=False)